pkg/ottl/parser_collection.go

```diff
@@ -0,0 +1,289 @@
+// so is can be returned by the ParserCollection.ParseStatements and ParserCollection.ParseStatementsWithContext
+// functions.
+type ParsedStatementConverter[T any, S StatementsGetter, R any] func(
+	collection *ParserCollection[S, R],
+	parser *Parser[T],
+	context string,
+	statements S,
+	parsedStatements []*Statement[T],
+) (R, error)
+
+func newNopParsedStatementConverter[T any, S StatementsGetter]() ParsedStatementConverter[T, S, any] {
+	return func(
+		_ *ParserCollection[S, any],
+		_ *Parser[T],
+		_ string,
+		_ S,
+		parsedStatements []*Statement[T]) (any, error) {
+		return parsedStatements, nil
+	}
+}
+
+// WithParserCollectionContext configures an ottl.Parser for the given context.
+// The provided ottl.Parser must be configured to support the provided context using
+// the ottl.WithPathContextNames option.
+func WithParserCollectionContext[K any, S StatementsGetter, R any](
+	context string,
+	parser *Parser[K],
+	converter ParsedStatementConverter[K, S, R],
+) ParserCollectionOption[S, R] {
+	return func(mp *ParserCollection[S, R]) error {
+		if _, ok := parser.pathContextNames[context]; !ok {
+			return fmt.Errorf(`context "%s" must be a valid "%T" path context name`, context, parser)
+		}
+		mp.contextParsers[context] = &parserCollectionParser[S]{
+			ottlParser:          newParserWrapper[K, S](parser),
+			statementsConverter: newStatementsConverterWrapper(converter),
+		}
+		return nil
+	}
+}
+
+// WithParserCollectionErrorMode has no effect on the ParserCollection, but might be used
+// by the ParsedStatementConverter functions to handle/create StatementSequence.
+func WithParserCollectionErrorMode[S StatementsGetter, R any](errorMode ErrorMode) ParserCollectionOption[S, R] {
+	return func(tp *ParserCollection[S, R]) error {
+		tp.ErrorMode = errorMode
+		return nil
+	}
+}
+
+// EnableParserCollectionModifiedStatementLogging controls the statements modification logs.
+// When enabled, it logs any statements modifications performed by the parsing operations,
+// instructing users to rewrite the statements accordingly.
+func EnableParserCollectionModifiedStatementLogging[S StatementsGetter, R any](enabled bool) ParserCollectionOption[S, R] {
+	return func(tp *ParserCollection[S, R]) error {
+		tp.modifiedStatementLogging = enabled
+		return nil
+	}
+}
+
+// ParseStatements parses the given statements into [R] using the configured context's ottl.Parser
+// and subsequently calling the ParsedStatementConverter function.
+// The statement's context is automatically inferred from the [Path.Context] values, choosing the
+// highest priority context found.
+// If no contexts are present in the statements, or if the inferred value is not supported by
+// the [ParserCollection], it returns an error.
+// If parsing the statements fails, it returns the underline [ottl.Parser.ParseStatements] error.
+func (pc *ParserCollection[S, R]) ParseStatements(statements S) (R, error) {
```

Comment:
Instead of taking an interface can this take `[]string` like the parser's ParseStatements?

Reply:
That'd possible, but the idea of making it an interface was giving the`ParsedStatementConverter` functions more flexibility to receive extra data, without the need of adding an extra argument for some custom data.

For example, the `transformprocessor` builds the parser collection and them invokes the `ParseContextStatements` [function](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3c37b131871d76a21552b6b179d66a52831bba67/processor/transformprocessor/internal/common/logs.go#L107) passing on a custom type `ContextStatements` argument. Besides the `Statements []string` property, that type has a `Conditions []string` property used to convert ([see](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3c37b131871d76a21552b6b179d66a52831bba67/processor/transformprocessor/internal/common/logs.go#L71)) the parsed `ottl.Statement[K]` into `[R]`, which is only known after the parser collection is built.

In summary, having it as an interface allows the `ParsedStatementConverter` functions to access customised `ParseStatements` input arguments, avoiding extra parameters on the `ParsedStatementConverter` and making the parser collection reusable. 

WDYT?

Reply:
If it can at least be updated to use `StatementsGetter` as the argument instead of a generic that would be good.